### PR TITLE
Add married-combined logic to the MS income tax

### LINKED ISSUE #3136: Add married-combined logic to the MS income tax
In the 0.497.2 version, PolicyEngine-US includes the Mississippi **married-joint** tax-calculation logic, but has not yet added the **married-combined** tax-calculation logic.

Here is how the [2021 MS income tax instructions](https://www.dor.ms.gov/sites/default/files/Forms/Individual/80100211_0.pdf#page=5) describe those two different methods of tax calculation:

```
Filing Status for Married Persons

Married persons may file tax returns in any of these three methods: 
1) joint, 2) combined or 3) separate.
Choose the method which results in the least amount of tax.

1) A joint return is usually completed when only one spouse has income.
    Place all income, deductions, exemptions, etc. in Column A (Taxpayer).

2) A combined return is completed when both spouses have income.
    Place one spouse's income in Column A (Taxpayer) and the other spouse's
    income in Column B (Spouse). The exemptions and deductions may be 
    divided in any manner you choose.  If only one spouse has income, this 
    income may not be split between husband and wife.
```

It is important to add the **married-combined** logic because TAXSIM35 includes that logic.   But TAXSIM35 does not include the third married-separate logic for the obvious reasons.

### PR DESCRIPTION
**Pull request by ZiqianCui**
Fixes #3136

### CODE REVIEW

File: policyengine_us/tests/policy/baseline/gov/states/ms/tax/income/deductions/ms_deductions_indiv.yaml
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/ms/tax/income/deductions/ms_deductions_indiv.yaml:1]: can we create an integration test with head and spouse income and the inputs from the tax form with a number of outputs to test our logic

File: policyengine_us/variables/gov/states/ms/tax/income/deductions/standard/ms_standard_deduction_indiv.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+                fsvals.SINGLE,
+                fsvals.SEPARATE,
+                fsvals.HEAD_OF_HOUSEHOLD,
+                fsvals.WIDOW,
+            ],
+        )
+        is_head = person("is_tax_unit_head", period)
+        is_spouse = person("is_tax_unit_spouse", period)
+        p = parameters(period).gov.states.ms.tax.income.deductions.standard
+        return (is_head | is_spouse) * p.amount[filing_status]
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/deductions/standard/ms_standard_deduction_indiv.py:36]:
```suggestion
        head_or_spouse = person("is_tax_unit_head_or_spouse", period)
        p = parameters(period).gov.states.ms.tax.income.deductions.standard
        return head_or_spouse * p.amount[filing_status]
```

File: policyengine_us/variables/gov/states/ms/tax/income/ms_taxable_income_joint.py [hunk 2]
```diff
@@ -13,8 +13,8 @@ class ms_taxable_income(Variable):
     )
     defined_for = StateCode.MS
 
-    def formula(tax_unit, period, parameters):
-        agi = tax_unit("ms_agi", period)
-        deductions = tax_unit("ms_deductions", period)
-        exemptions = tax_unit("ms_total_exemptions", period)
+    def formula(person, period, parameters):
+        agi = person.tax_unit("ms_agi", period)
```
Comment (Xinyi-Shao) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_taxable_income_joint.py:17]: Shall it be person.tax_unit.sum?

File: policyengine_us/variables/gov/states/ms/tax/income/ms_taxable_income_joint.py [hunk 3]
```diff
@@ -13,8 +13,8 @@ class ms_taxable_income(Variable):
     )
     defined_for = StateCode.MS
 
-    def formula(tax_unit, period, parameters):
-        agi = tax_unit("ms_agi", period)
-        deductions = tax_unit("ms_deductions", period)
```
Comment (Xinyi-Shao) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_taxable_income_joint.py:18]: Unsure if it needs a for statement

File: policyengine_us/variables/gov/states/ms/tax/income/deductions/ms_deductions_indiv.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+class ms_deductions_indiv(Variable):
+    value_type = float
+    entity = Person
+    label = "Mississippi deductions filing seperately"
+    unit = USD
+    definition_period = YEAR
+    reference = (
+        "https://www.dor.ms.gov/sites/default/files/Forms/Individual/80108228.pdf#page=1"
+        "https://www.law.cornell.edu/regulations/mississippi/35-Miss-Code-R-SS-3-02-11-103"
```
Comment (Xinyi-Shao) [inline policyengine_us/variables/gov/states/ms/tax/income/deductions/ms_deductions_indiv.py:12]: May add more source that directly explains an individual should claim itemized deductions or standard deduction, whichever is larger

**Comment by PavelMakarchuk** [comment]
@MaxGhenis would we have to branch for the following married combined logic. The Standard deduction and exemptions can be divided between head and spouse as the filers want (minimizing taxable income). Or would prorating it and making sure there is no "waste" if possible be sufficient?

image

**Comment by MaxGhenis** [comment]
For now let's assume the spouse with the higher agi claims the deductions. IIRC @nikhilwoodruff  did something more sophisticated in another program, Nikhil do you remember?

**Comment by PavelMakarchuk** [comment]
> For now let's assume the spouse with the higher agi claims the deductions. IIRC @nikhilwoodruff did something more sophisticated in another program, Nikhil do you remember?

this is part of the married combined logic, however it differs from Martins approach in IA, different rules

File: policyengine_us/variables/gov/states/ms/tax/income/deductions/ms_deductions_joint.py [hunk 5]
```diff
@@ -1,10 +1,10 @@
 from policyengine_us.model_api import *
 
 
-class ms_deductions(Variable):
+class ms_deductions_joint(Variable):
     value_type = float
     entity = TaxUnit
-    label = "Mississippi deductions"
+    label = "Mississippi deductions filing jointly"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/deductions/ms_deductions_joint.py:7]:
```suggestion
    label = "Mississippi deductions when married couples file jointly"
```

File: policyengine_us/variables/gov/states/ms/tax/income/deductions/standard/ms_standard_deduction_indiv.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
+from policyengine_us.model_api import *
+
+
+class ms_standard_deduction_indiv(Variable):
+    value_type = float
+    entity = Person
+    label = "Mississippi personal standard deduction when married couples file separately"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/deductions/standard/ms_standard_deduction_indiv.py:7]:
```suggestion
    label = "Mississippi standard deduction when married couples file separately"
```

File: policyengine_us/variables/gov/states/ms/tax/income/exemptions/ms_total_exemptions_indiv.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
+from policyengine_us.model_api import *
+
+
+class ms_total_exemptions_indiv(Variable):
+    value_type = float
+    entity = Person
+    label = "Mississippi total exemptions"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/exemptions/ms_total_exemptions_indiv.py:7]:
```suggestion
    label = "Mississippi total exemptions when married couples file separately"
```

File: policyengine_us/variables/gov/states/ms/tax/income/exemptions/ms_total_exemptions_joint.py [hunk 8]
```diff
@@ -1,7 +1,7 @@
 from policyengine_us.model_api import *
 
 
-class ms_total_exemptions(Variable):
+class ms_total_exemptions_joint(Variable):
     value_type = float
     entity = TaxUnit
     label = "Mississippi total exemptions"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/exemptions/ms_total_exemptions_joint.py:7]:
```suggestion
    label = "Mississippi total exemptions hen married couples file jointly"
```

File: policyengine_us/variables/gov/states/ms/tax/income/exemptions/ms_total_exemptions_indiv.py [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+    unit = USD
+    definition_period = YEAR
+    defined_for = StateCode.MS
+    reference = "https://www.dor.ms.gov/sites/default/files/Forms/Individual/80105228.pdf#page=1"
+
+    def formula(person, period, parameters):
+        head_or_spouse = person("is_tax_unit_head_or_spouse", period)
+        total_exemptions = person.tax_unit("ms_total_exemptions_joint", period)
+        # Per the atx form, the exemption amount is split in half between the head
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/exemptions/ms_total_exemptions_indiv.py:16]:
```suggestion
        # Per the tax form, the exemption amount is split in half between the head
```

File: policyengine_us/variables/gov/states/ms/tax/income/ms_agi_indiv.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
+from policyengine_us.model_api import *
+
+
+class ms_agi_indiv(Variable):
+    value_type = float
+    entity = Person
+    label = "Mississippi adjusted gross income"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_agi_indiv.py:7]:
```suggestion
    label = "Mississippi adjusted gross income when married couples file separately"
```

File: policyengine_us/variables/gov/states/ms/tax/income/ms_agi_indiv.py [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+    )
+    defined_for = StateCode.MS
+
+    # AGI = Income - Total adjustments from gross income
+    adds = "gov.states.ms.tax.income.income_sources"
+    subtracts = ["ms_agi_adjustments"]
+
+    def formula(person, period, parameters):
+        head_or_spouse = person("is_tax_unit_head_or_spouse", period)
+        ms_agi = person.tax_unit("ms_agi_joint", period)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_agi_indiv.py:22]:
```suggestion
        p = parameters(period).gov.states.ms.tax.income
        income_sources = add(person, period, p.income_sources)
        adjustments = person("ms_agi_adjustments", period)
        adjusted_income = max_(0, income_sources - adjustments)
        return head_or_spouse * adjusted_income
```

File: policyengine_us/variables/gov/states/ms/tax/income/ms_agi_indiv.py [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+        # Per the atx form, the exemption amount is split in half between the head
+        # and the spouse of the household
+        return head_or_spouse * (0.5 * ms_agi)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_agi_indiv.py:25]:
```suggestion

```

File: policyengine_us/variables/gov/states/ms/tax/income/ms_agi_joint.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_agi_joint.py:1]: can AGI be negative? if not we need to change this to a formula with max_

File: policyengine_us/variables/gov/states/ms/tax/income/ms_files_separately.py [hunk 13]
```diff
@@ -0,0 +1,15 @@
+from policyengine_us.model_api import *
+
+
+class ms_files_separately(Variable):
+    value_type = bool
+    entity = TaxUnit
+    label = "married couple files separately on Mississippi tax return"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_files_separately.py:7]:
```suggestion
    label = "Married couple filing separately on the mississippi tax return"
```

File: policyengine_us/variables/gov/states/ms/tax/income/ms_files_separately.py [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+    definition_period = YEAR
+    reference = ()
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_files_separately.py:9]: add

File: policyengine_us/variables/gov/states/ms/tax/income/ms_income_tax.py [hunk 15]
```diff
@@ -4,7 +4,7 @@
 class ms_income_tax(Variable):
     value_type = float
     entity = TaxUnit
-    label = "Mississippi income tax"
+    label = "Mississippi income tax filing seperately"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_income_tax.py:7]:
```suggestion
    label = "Mississippi income tax when married couples file separately"
```

File: policyengine_us/variables/gov/states/ms/tax/income/ms_income_tax_before_credits.py [hunk 16]
```diff
@@ -10,6 +10,14 @@ class ms_income_tax_before_credits(Variable):
     defined_for = StateCode.MS
 
     def formula(tax_unit, period, parameters):
-        income = tax_unit("ms_taxable_income", period)
+        filing_separately = tax_unit("ms_files_separately", period)
+        income_separate = add(tax_unit, period, ["ms_taxable_income_indiv"])
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_income_tax_before_credits.py:14]:
```suggestion
        income_filing_separately = add(tax_unit, period, ["ms_taxable_income_indiv"])
```

File: policyengine_us/variables/gov/states/ms/tax/income/ms_income_tax_before_credits.py [hunk 17]
```diff
@@ -10,6 +10,14 @@ class ms_income_tax_before_credits(Variable):
...
+        income_joint = tax_unit("ms_taxable_income_joint", period)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_income_tax_before_credits.py:15]:
```suggestion
        income_filing_jointly = tax_unit("ms_taxable_income_joint", period)
```

File: policyengine_us/variables/gov/states/ms/tax/income/ms_taxable_income_indiv.py [hunk 18]
```diff
@@ -0,0 +1,20 @@
+from policyengine_us.model_api import *
+
+
+class ms_taxable_income_indiv(Variable):
+    value_type = float
+    entity = Person
+    label = "Mississippi taxable income individually"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_taxable_income_indiv.py:7]:
```suggestion
    label = "Mississippi taxable income when married couples file separately"
```

**Review by PavelMakarchuk — CHANGES_REQUESTED** [review]

File: policyengine_us/variables/gov/states/ms/tax/income/ms_agi_indiv.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_agi_indiv.py:1]: we need to have a structure like [this](https://github.com/PolicyEngine/policyengine-us/pull/3516/files#diff-021f7f80aeb0e2208b18a44cad6c89f0811f1dca928f97d3ecc64c438a4fdd04:~:text=policyengine_us/variables/gov/states/ar/tax/income/ar_agi.py) for the AGI

File: policyengine_us/variables/gov/states/ms/tax/income/ms_income_tax.py [hunk 19]
```diff
@@ -4,7 +4,7 @@
 class ms_income_tax(Variable):
     value_type = float
     entity = TaxUnit
-    label = "Mississippi income tax"
+    label = "Mississippi income tax filing seperately when married couples file separately"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_income_tax.py:7]: no need for this change

File: policyengine_us/variables/gov/states/ms/tax/income/ms_agi_indiv.py
Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/ms/tax/income/ms_agi_indiv.py:1]: we just make the agi person level but we can keep one file where we attribute the dependent income to the head 
image

File: policyengine_us/variables/gov/states/ms/tax/income/ms_taxable_income_joint.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_taxable_income_joint.py:1]: can you add the following files:

state_filing_status_if_married_filing_separately_on_same_return

ms_can_file_separate_on_same_return

[this](https://github.com/PolicyEngine/policyengine-us/pull/3516/files#diff-9e807e0d8b3996ed47a0026c5abe5d2e9ad2eb416282bf9b82a24c37bcb2c686) PR can be an example

**Comment by PavelMakarchuk** [comment]
Duplicate PR

**Pull request closed without merging**